drivers/DeviceKey.cpp

```diff
@@ -0,0 +1,255 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "drivers/DeviceKey.h"
+#include "mbedtls/config.h"
+#include "mbedtls/cmac.h"
+#include "nvstore.h"
+#include "trng_api.h"
+
+#if !defined(MBEDTLS_CMAC_C)
+#error [NOT_SUPPORTED] MBEDTLS_CMAC_C needs to be enabled for this driver
+#else
+
+#if NVSTORE_ENABLED
+
+namespace mbed {
+
+DeviceKey::DeviceKey()
+{
+    return;
+}
+
+DeviceKey::~DeviceKey()
+{
+    return;
+}
+
+int DeviceKey::generate_derived_key(const unsigned char *salt, size_t isalt_size, unsigned char *output,
+                                    uint16_t ikey_type)
+{
+    uint32_t key_buff[DEVICE_KEY_32BYTE / sizeof(uint32_t)];
+    size_t actual_size = DEVICE_KEY_32BYTE;
+
+    if (DEVICE_KEY_16BYTE != ikey_type && DEVICE_KEY_32BYTE != ikey_type) {
+        return DEVICEKEY_INVALID_KEY_TYPE;
+    }
+
+    //First try to read the key from NVStore
+    int ret = read_key_from_nvstore(key_buff, actual_size);
+    if (DEVICEKEY_SUCCESS != ret && DEVICEKEY_NOT_FOUND != ret) {
+        return ret;
+    }
+
+    if (DEVICE_KEY_16BYTE != actual_size && DEVICE_KEY_32BYTE != actual_size) {
+        return DEVICEKEY_READ_FAILED;
+    }
+
+    //If the key was not found in NVStore we will create it by using TRNG and then save it to NVStore
+    if (DEVICEKEY_NOT_FOUND == ret) {
+        ret = generate_key_by_trng(key_buff, actual_size);
+        if (DEVICEKEY_SUCCESS != ret) {
+            return ret;
+        }
+
+        ret = device_inject_root_of_trust(key_buff, actual_size);
+        if (DEVICEKEY_SUCCESS != ret) {
+            return ret;
+        }
+    }
+
+    ret = get_derived_key(key_buff, actual_size, salt, isalt_size, output, ikey_type);
+    return ret;
+}
+
+int DeviceKey::device_inject_root_of_trust(uint32_t *value, size_t isize)
+{
+    return write_key_to_nvstore(value, isize);
+}
+
+int DeviceKey::write_key_to_nvstore(uint32_t *input, size_t isize)
+{
+    if (DEVICE_KEY_16BYTE != isize && DEVICE_KEY_32BYTE != isize) {
+        return DEVICEKEY_INVALID_KEY_SIZE;
+    }
+
+    //First we read if key exist. If it is exists, we return DEVICEKEY_ALREADY_EXIST error
+    uint32_t read_key[DEVICE_KEY_32BYTE / sizeof(uint32_t)] = {0};
```

Comment:
Why is `read_key` not defined as a `uint8_t` array?

Reply:
@hanno-arm This is because NVStore had API of uint32_t. This is internal allocation therefore has no impact externally. Do you have any concerns?

Reply:
@dannybenor No, I don't have concerns, I only found it surprising when I read it.